finality-aleph/src/justification/requester.rs

```diff
@@ -202,6 +182,55 @@ where
     }
 
     pub fn finalized_number(&self) -> NumberFor<B> {
-        self.client.info().finalized_number
+        self.backend.blockchain().info().finalized_number
+    }
+
+    fn request(&mut self, hash: <B as BlockT>::Hash) {
+        if let Ok(Some(header)) = self.backend.blockchain().header(BlockId::Hash(hash)) {
+            let number = *header.number();
+            debug!(target: "aleph-justification", "Trying to request block {:?}", number);
+            self.request_status.save_block_number(number);
+            self.request_status.insert_hash(hash);
+            debug!(target: "aleph-justification", "We have block {:?} with hash {:?}. Requesting justification.", number, header.hash());
+            self.justification_request_scheduler.on_request_sent();
+            self.block_requester.request_justification(&hash, number);
+        } else {
+            debug!(target: "aleph-justification", "Cancelling request, because we don't have block {:?}.", hash);
+        }
+    }
+
+    // We request justifications for all the children of last finalized block and a justification
+    // for a block of number num on longest branch.
+    // Assuming that we request at the same pace that finalization is progressing, the former ensures
+    // that we are up to date with finalization. On the other hand, the former enables fast catch up
+    // if we are behind.
+    // We don't remove the child that it's on the same branch as best since a fork may happen
+    // somewhere in between them.
+    fn request_targets(&self, mut num: NumberFor<B>) -> Vec<<B as BlockT>::Hash> {
+        let blockchain_backend = self.backend.blockchain();
+        let blockchain_info = blockchain_backend.info();
+        let finalized_hash = blockchain_info.finalized_hash;
+
+        let mut targets = blockchain_backend
+            .children(finalized_hash)
+            .unwrap_or_default();
+        let best_number = blockchain_info.best_number;
+        if best_number < num {
+            // most probably block best_number is not yet finalized
+            num = best_number - 1;
+        }
+        match blockchain_backend.hash(num) {
+            Ok(Some(hash)) => {
+                targets.push(hash);
+            }
+            Ok(None) => {
+                debug!(target: "aleph-justification", "Cancelling request, because we don't have block {:?}.", num);
+            }
+            Err(err) => {
+                debug!(target: "aleph-justification", "Cancelling request, because fetching block {:?} failed {:?}.", num, err);
```

Comment:
This and the one above sound like quite unexpected occurrences – the backend claims we have a block with this number but refuses to show it? Mighty suspicious... I would `warn` in both these situations.

Reply:
changed.